src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java

```diff
@@ -404,7 +415,70 @@ protected AbstractSimulation (Builder builder) {
+            beaconsObservedCurrentInterval.remove(beacon);
+        }
+
+        if (!beaconsObservedCurrentInterval.containsKey(beacon)) {
+            beaconsObservedCurrentInterval.put(beacon, new AutoValue_ObservedInterval.Builder()
+                                                                                  .setStart(currentRound)
+                                                                                  .setObserved(observed));
+        }
+        beaconsObservedCurrentInterval.get(beacon).setEnd(currentRound);
+    }
+
+    private void closeBeaconsOpenObservedIntervals() {
+        for (Beacon beacon : beaconsObservedCurrentInterval.keySet()) {
+            ObservedInterval.Builder currentInterval = beaconsObservedCurrentInterval.get(beacon);
+            beaconsObservedIntervals.computeIfAbsent(beacon, b -> new ArrayList<>()).add(currentInterval.build());
+        }
+        beaconsObservedCurrentInterval.clear();
+    }
+
     private static double distance(Location firstLocation, Location secondLocation) {
         return Math.abs(firstLocation.row() - secondLocation.row()) + Math.abs(firstLocation.col() - secondLocation.col());
     }
+
+    private Table<String, String, Double> calculateObservedStats(Map<String, List<Integer>> observedIntervals,
+                                                                 Map<String, List<Integer>> unobservedIntervals) {
+        ImmutableTable.Builder<String, String, Double> observedStats = new ImmutableTable.Builder<>();
+        mapIdsToPercentOfSimulation(observedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.observedPercent, value));
+        mapIdsToMinValue(observedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.minimumLengthObservedInterval, value));
+        mapIdsToMinValue(unobservedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.minimumLengthUnobservedInterval, value));
+        mapIdsToMaxValue(observedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.maximumLengthObservedInterval, value));
+        mapIdsToMaxValue(unobservedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.maximumLengthUnobservedInterval, value));
+        mapIdsToAvgValue(observedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.averageLengthObservedInterval, value));
+        mapIdsToAvgValue(unobservedIntervals).forEach((beaconId, value) ->
+                observedStats.put(beaconId, Schema.StatisticsState.averageLengthUnobservedInterval, value));
+        return observedStats.build();
+    }
+
+    private Map<String, Double> mapIdsToPercentOfSimulation(Map<String, List<Integer>> idsToIntervals) {
```

Comment:
This looks like the same code as in line 158 - right? If so, (a) consider applying the same simplification I recommended there; (b) consider using this method from line 158, avoiding duplication.